raiden-ts/src/transfers/epics/locked.ts

```diff
@@ -492,38 +805,49 @@ export const transferGenerateAndSignEnvelopeMessageEpic = (
   action$: Observable<RaidenAction>,
   {}: Observable<RaidenState>,
   deps: RaidenEpicDeps,
-): Observable<
-  | transferSigned
-  | transferSecret
-  | transferUnlock.success
-  | transferUnlock.failure
-  | transfer.failure
-  | transferExpire.success
-  | transferExpire.failure
-  | withdrawReceive.success
-> => {
+) => {
   const withdrawCache = new LruCache<string, Signed<WithdrawConfirmation>>(32);
-  const latestState$ = deps.latest$.pipe(pluckDistinct('state'));
-  return action$.pipe(
-    filter(
-      isActionOf([
-        transfer.request,
-        transferUnlock.request,
-        transferExpire.request,
-        withdrawReceive.request,
-      ]),
+  const state$ = deps.latest$.pipe(pluckDistinct('state')); // replayed(1)' state$
+  return merge(
+    action$.pipe(
+      filter(
+        isActionOf([
+          transfer.request,
+          transferUnlock.request,
+          transferExpire.request,
+          withdrawReceive.request,
+        ]),
+      ),
+    ),
+    // merge separatedly, to filter per message type before concat
+    action$.pipe(
+      filter(
+        isMessageReceivedOfType([Signed(LockedTransfer), Signed(Unlock), Signed(LockExpired)]),
+      ),
+    ),
+  ).pipe(
+    concatMap(action =>
+      transfer.request.is(action)
```

Comment:
Just out of curiosity, was there a problem with the switch?

Reply:
Yes, unfortunately: as we need to `switch` with `messageReceivedTyped`, and for the same action type (`messageReceived`) switch different functions (receive LockedTransfer, Unlock, LockExpire) under the same lock, and switch can only compare the value of the type, we couldn't use it for a shared action narrowing on the types of a deep property (i.e. `payload.message.type`). An intermediary option, but which would make for heterogenous handling, is to `switch` on the action type `messageReceived`, and nested, `if` or `switch` on the message type; I prefered the former, but let me know if you guys would prefer the later, I don't have strong feelings about this.

Reply:
Ok so we don't only narrow in a single condition but in some case we do on multiple now.

Reply:
Yes. We could have simple different epics for handling  `messageReceived`, filtering for each message and outputting specific actions for each, but as there's no logic in such epics, I'd say it'd not be worth the hassle. We can go this way in the future if needed, but now, should be enough.

Reply:
> We could have simple different epics for handling `messageReceived`, filtering for each message and outputting specific actions for each,

Would be good for readability